akka-typed/src/main/scala/akka/typed/cluster/sharding/ClusterSharding.scala

```diff
@@ -0,0 +1,323 @@
+ * and the hashcode of the entityId to decide which shard an entity belongs to.
+ *
+ * This is recommended since it does not force details about sharding into the entity protocol
+ *
+ * @tparam A The type of message accepted by the entity actor
+ */
+final class HashCodeMessageExtractor[A](maxNumberOfShards: Int) extends ShardingMessageExtractor[ShardingEnvelope[A], A] {
+  def entityId(envelope: ShardingEnvelope[A]): String = envelope.entityId
+  def entityMessage(envelope: ShardingEnvelope[A]): A = envelope.message
+  def shardId(envelope: ShardingEnvelope[A]): String = (math.abs(envelope.entityId.hashCode) % maxNumberOfShards).toString
+}
+
+/**
+ * Java API:
+ *
+ * Default message extractor type, using a property of the message to identify what entity a message is for
+ * and the hashcode of the entityId to decide which shard an entity belongs to.
+ *
+ * This is recommended since it does not force details about sharding into the entity protocol
+ *
+ * @tparam A The type of message accepted by the entity actor
+ */
+abstract class HashCodeNoEnvelopeMessageExtractor[A](maxNumberOfShards: Int) extends ShardingMessageExtractor[A, A] {
+  final def entityMessage(message: A): A = message
+  def shardId(message: A): String = {
+    val id = entityId(message)
+    if (id != null) (math.abs(id.hashCode) % maxNumberOfShards).toString
+    else null
+  }
+
+  override def toString = s"HashCodeNoEnvelopeMessageExtractor($maxNumberOfShards)"
+}
+
+object ClusterSharding extends ExtensionId[ClusterSharding] {
+
+  override def createExtension(system: ActorSystem[_]): ClusterSharding =
+    new AdaptedClusterShardingImpl(system)
+
+  /** Java API */
+  def get(system: ActorSystem[_]): ClusterSharding = apply(system)
+}
+
+final class AdaptedClusterShardingImpl(system: ActorSystem[_]) extends ClusterSharding {
+  require(system.isInstanceOf[ActorSystemAdapter[_]], "only adapted untyped actor systems can be used for cluster features")
+
+  private val cluster = Cluster(system)
+  private val untypedSystem = ActorSystemAdapter.toUntyped(system)
+  private val untypedSharding = akka.cluster.sharding.ClusterSharding(untypedSystem)
+
+  override def spawn[A](
+    behavior:           Behavior[A],
+    entityProps:        Props,
+    typeName:           String,
+    settings:           ClusterShardingSettings,
+    maxNumberOfShards:  Int,
+    handOffStopMessage: A): ActorRef[ShardingEnvelope[A]] = {
+    val extractor = new HashCodeMessageExtractor[A](maxNumberOfShards)
+    spawn(behavior, entityProps, typeName, settings, extractor, defaultShardAllocationStrategy(settings), handOffStopMessage)
+  }
+
+  override def spawn[E, A](
+    behavior:           Behavior[A],
+    entityProps:        Props,
+    typeName:           String,
+    settings:           ClusterShardingSettings,
+    messageExtractor:   ShardingMessageExtractor[E, A],
+    handOffStopMessage: A): ActorRef[E] =
```

Comment:
so this is for the convenient defaults, and the other is the full. Then I think `entityProps` should not be included in this one.